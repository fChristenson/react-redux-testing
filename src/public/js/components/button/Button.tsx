import React from "react";

export class Button extends React.Component {
  render() {
    return <button {...this.props}>{this.props.children}</button>;
  }
}
