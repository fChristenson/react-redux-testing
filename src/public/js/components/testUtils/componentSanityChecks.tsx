import React from "react";
import { shallow } from "enzyme";

// The sanity checks, these are checks that verify that we haven't made a mistake and that our tools work
export const componentSanityChecks = (
  label: string,
  Element: React.ReactElement
) =>
  describe(`Sanity checks for ${label}`, () => {
    it("is defined", () => {
      expect(Element).toBeDefined();
    });

    it("renders the element", () => {
      const wrapper = shallow(Element);
      const expected = false;
      const actual = wrapper.isEmptyRender();
      expect(expected).toEqual(actual);
    });

    it("matches the snapshot", () => {
      const wrapper = shallow(Element);
      const expected = wrapper.html();
      expect(expected).toMatchSnapshot();
    });
  });
