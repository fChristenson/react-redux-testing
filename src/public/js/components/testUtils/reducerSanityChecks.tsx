import { action } from "../../../../lib/action";

// The sanity checks, these are checks that verify that we haven't made a mistake and that our tools work
export const reducerSanityChecks = (
  label: string,
  reducer: (state: any, action: any) => any
) =>
  describe(`Sanity checks for ${label}`, () => {
    it("is defined", () => {
      expect(reducer).toBeDefined();
    });

    it("returns the same state as passed in if no action matches", () => {
      const expected = { foo: 1 };
      const actual = reducer({ foo: 1 }, action("UNMATCHED_ACTION", null));
      expect(expected).toEqual(actual);
    });

    it("returns a initial state if none is passed in", () => {
      const actual = reducer(undefined, action("UNMATCHED_ACTION", null));
      expect(actual).toBeDefined();
      expect(typeof actual).toBe("object");
      expect(actual).not.toBe(null);
      expect(actual).not.toBe(undefined);
      expect(Array.isArray(actual)).toBe(false);
    });
  });
