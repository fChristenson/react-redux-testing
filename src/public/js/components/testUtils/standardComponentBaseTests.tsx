import React from "react";
import { shallow } from "enzyme";

/**
 * The common tests, these are tests that verify that a standard component has the intended behavior we want.
 * These tests are not specific to any one component and are meant to check the basic behavior.
 *
 * If we want we can reuse these sorts of test for standard components that fall in a certain group of element.
 */
export const standardComponentBaseTests = (Component: React.ComponentClass) =>
  describe("Standard component base tests", () => {
    it("should pass through props to the component", () => {
      const customAttributeName = `data-test${Math.random()}`;
      const props = {
        onClick: jest.fn(),
        [customAttributeName]: "foo"
      };
      // @ts-ignore
      const wrapper = shallow(<Component {...props} />);
      wrapper.simulate("click");
      expect(props.onClick).toBeCalledTimes(1);
      // @ts-ignore
      expect(wrapper.first().props()[customAttributeName]).toEqual("foo");
    });
  });
