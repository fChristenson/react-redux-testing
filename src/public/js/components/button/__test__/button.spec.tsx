import React from "react";
import { Button } from "../Button";
import { shallow } from "enzyme";
import { standardComponentBaseTests } from "../../testUtils/standardComponentBaseTests";
import { componentSanityChecks } from "../../testUtils/componentSanityChecks";

/**
 * A button is what we want to think of as a standard component, in other words it is a component that we will
 * find in many different places in our codebase and this makes it very important to test.
 *
 * The reason is simple, if you want to change something about the button how do you know you didn't break
 * something in one of the places it is being used?
 */
describe("Button", () => {
  componentSanityChecks("Button", <Button />);
  standardComponentBaseTests(Button);

  // component tests, these are the tests that are specific to this component
  it("should set the text of the button", () => {
    const wrapper = shallow(<Button>Hello</Button>);
    expect(wrapper.first().text()).toEqual("Hello");
  });
});
