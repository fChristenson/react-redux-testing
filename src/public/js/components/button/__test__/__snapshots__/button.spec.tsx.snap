// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Sanity checks for Button matches the snapshot 1`] = `"<button></button>"`;
